### ISSUE TITLE
DNS server coming from vpn network is not reflected in WSL

### ISSUE THREAD

**Issue by asvetliakov** [Microsoft/BashOnWindows#1350]
* A brief description
I've L2TP/IPsec vpn connection without default gateway set and own DNS server

* Expected results
Bash should add VPN DNS IP to /etc/resolv.conf

* Actual results (with terminal output if applicable)
No VPN DNS IP in /etc/resolve.conf . It works though if i set "use default gateway on remote network" (generally i don't want) setting in vpn configuration.

* Your Windows build number
14965.1001

![2016-11-12](https://cloud.githubusercontent.com/assets/8881674/20236046/c006e79c-a8a6-11e6-84ee-0d803d622943.png)

**Comment by sunilmut** [comment]
@asvetliakov - Thanks for reporting the issue. Yes, we are aware of issues with VPN, as you can also see in #416. We are actively working on a solution for this. @misenesi as FYI.

**Comment by misenesi** [comment]
Hi @asvetliakov, could you please provide output when you do:

ipconfig /all

?

I have a fix prepared for this, but need to verify that your VPN networking interface is reported as point-to-point interface.

**Comment by asvetliakov** [comment]
@misenesi 
```
PS C:\Users\asvet> ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : Alexey-PC
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : Home

Ethernet adapter Ethernet:

   Connection-specific DNS Suffix  . : Home
   Description . . . . . . . . . . . : Intel(R) Ethernet Connection I217-V
   Physical Address. . . . . . . . . : E0-3F-49-AC-04-1E
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::5867:7e40:cdf2:456a%25(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.1.33(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : Wednesday, November 30, 2016 2:28:47 AM
   Lease Expires . . . . . . . . . . : Wednesday, November 30, 2016 2:28:47 PM
   Default Gateway . . . . . . . . . : 192.168.1.1
   DHCP Server . . . . . . . . . . . : 192.168.1.1
   DHCPv6 IAID . . . . . . . . . . . : 65027913
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1E-92-04-D1-E0-3F-49-AC-04-1E
   DNS Servers . . . . . . . . . . . : 80.58.61.250
                                       80.58.61.254
   NetBIOS over Tcpip. . . . . . . . : Enabled

PPP adapter freechat.com:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : freechat.com
   Physical Address. . . . . . . . . :
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 10.61.0.5(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.255
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : 10.60.10.121
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Ethernet 12:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : TAP-Windows Adapter V9
   Physical Address. . . . . . . . . : 00-FF-2F-01-1A-D1
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Tunnel adapter Local Area Connection* 10:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Teredo Tunneling Adapter
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv6 Address. . . . . . . . . . . : 2001:0:9d38:6ab8:2cca:22a9:3f57:fede(Preferred)
   Link-local IPv6 Address . . . . . : fe80::2cca:22a9:3f57:fede%19(Preferred)
   Default Gateway . . . . . . . . . : ::
   DHCPv6 IAID . . . . . . . . . . . : 318767104
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1E-92-04-D1-E0-3F-49-AC-04-1E
   NetBIOS over Tcpip. . . . . . . . : Disabled
```
The one with name 'freechat.com' is my VPN

**Comment by misenesi** [comment]
Thank you @asvetliakov . I have checked in a fix that will address some of the VPN DNS resolving issues, including yours. Currently for DNS resolving we update /etc/resolv.conf from the service. In this fix I added the capability to manually modify the /etc/resolv.conf file if you wish so, disabling its automatic regeneration.

However, due to how DNS resolving works for various VPN solutions, this fix will only work with strict force tunnel VPN. I have a proposal for better fix prepared for that and we are currently discussing it.

**Comment by blemasle** [comment]
@misenesi With creators update installed, the automic generation cannot be disabled on my system. Even if I remove the first line, it still get regenerated every single time. It's working for /etc/hosts though.

**Comment by sunilmut** [comment]
@blemasle - @misenesi is no longer with MSFT.
It looks like we screwed up the `/etc/resolv.conf` auto-generation logic, see @benhillis [comment in #1908](https://github.com/Microsoft/BashOnWindows/issues/1908#issuecomment-294620663). 

Here is the background. By default, bash.exe will auto-generate `/etc/resolv.conf` every time you launch bash.exe. Then it will try to keep it up to date with changes from Windows, when bash is running. It looks like if you remove the first line, we stop auto-update of the file while bash is running, but forgot to disable the logic to auto-generate during bash launch. We do apologize for the inconvenience and as @benhillis mentions in #1908, we are looking to improve the experience here.

**Comment by johnarban** [comment]
I don't know if this is still a problem for anyone else. I am on build 16232.rs_prerelease.170624-1334 using WSL with Ubuntu 16.04.2 LTS from the Windows Store and CISCO AnyConnect version 21.12020 , and I still can't connect. I tried removing the commented line in `/etc/resolv.conf`, but it still get's reset. I don't know if there has a been a fix posted yet, but I haven't found it online if it exists.

**Comment by mo18** [comment]
The problem I'm seeing is that the ordering of the dns servers is incorrect when the vpn is connected. as a reslult, I can't resolve any of the hosts behind the vpn. I'm running windows 15063.483 with Cisco Anyconnect. The dns bindings in windows are the following:
```
> Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses
10.0.0.51
10.0.0.52
192.168.1.1
```

in ubuntu the are the following:
```
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, remove this line.
nameserver 192.168.1.1
nameserver 10.134.0.51
nameserver 10.134.0.52
search home xyz.vpn
```

**Comment by sunilmut** [comment]
If you are connected to a VPN and lose connectivity within bash, please try the workaround posted [here](https://msdn.microsoft.com/en-us/commandline/wsl/troubleshooting#bash-loses-network-connectivity-once-connected-to-a-vpn). It should work for Creators Update and above. Post Fall Creators Update, we will be looking at a better support for other VPN solutions.

Thanks to @bradley101, who first pointed out the workaround.

**Comment by johnarban** [comment]
Thank you that works around that. The only other thing is to get the list of domain suffixes, but that is 2nd order.

**Comment by nimbixler** [comment]
here's a more automatic workaround that works for me - in my case I only ever connect to one VPN at a time and its nameserver is at the very end of /etc/resolv.conf, which of course is no good.

YMMV but, first...

create a file in /etc/sudoers.d with allowing your username to run sudo with no password; you can restrict this to specific commands if you want but I do it for all (please adjust for your needs).  For example let's say your username is linux, so create the file /etc/sudoers.d/00-linux (you can call this whatever you want), with the following text in it:

```
linux ALL=(ALL) NOPASSWD: ALL
```

you'll need to of course sudo to root to create that file.  Be sure to replace "linux" with whatever your username is in WSL.

Close the shell and all instances of WSL and run a new one, and try it - type sudo -s
if you are not prompted for password, it worked.  Otherwise please check the syntax.

Next, create this python file in your home directory = call it resolv_flip.py :

```
#!/usr/bin/python

if __name__ == '__main__':
    with open('/etc/resolv.conf', 'r') as f:
        lines = f.read().splitlines()
    ns = []
    other = []
    for i in lines:
        if i[:11] != 'nameserver ':
            other.append(i)
        elif i[11:][:2] == '8.':
            ns.append(i)
        else:
            ns.insert(0, i)
    with open('/etc/resolv.conf', 'w') as f:
        f.write('\n'.join(other) + '\n')
        f.write('\n'.join(ns) + '\n')
```

finally, add this to the end of your ~/.bashrc file:

```
sudo ~/resolv_flip.py
```

from that point forward every time you run the WSL shell it will flip the nameservers so that anything that doesn't start with 8. will be at the top, but it won't disturb the comment header that is used to autogenerate the file (because we need that information in order for this to work).

It's not perfect but it works fine for me, and it's dead simple.

Enjoy,
Leo Reiter

**Comment by l8nite** [comment]
For me, DNS lookups via my corporate nameserver (on VPN) won't resolve external domains, so just swapping nameserers in resolv.conf only allows me to swap between "only resolve external" and "only resolve internal"

I'm still hoping something can be done to proxy DNS requests out to the windows host resolver.

**Comment by nimbixler** [comment]
Makes sense.. this worked for me but my DNS does external as well.  It's too bad this is isn't properly fixed yet.. it's a great set of tools that are almost unusable as a result.

**Comment by sunilmut** [comment]
@l8nite - If the VPN DNS server is not able to resolve a name, then glibc resolver should walk the list in `/etc/resolv.conf` and try the next nameserver on the list. Are you suggesting that that's not happening?

**Comment by nimbixler** [comment]
that's not how the resolver works on Linux... it only goes to the next server if the primary (or secondary, etc.) is not accessible.  Otherwise it would work fine to just keep it as it is.  In our case our DNS can do external resolution as well, so it's not a problem for me to just switch the order.

**Comment by nimbixler** [comment]
@sunilmut FYI, this explains it pretty well:
http://man7.org/linux/man-pages/man5/resolv.conf.5.html

specifically:
"nameserver Name server IP address
              Internet address of a name server that the resolver should
              query, either an IPv4 address (in dot notation), or an IPv6
              address in colon (and possibly dot) notation as per RFC 2373.
              Up to MAXNS (currently 3, see <resolv.h>) name servers may be
              listed, one per keyword.  If there are multiple servers, the
              resolver library queries them in the order listed.  If no
              nameserver entries are present, the default is to use the name
              server on the local machine.  (The algorithm used is to try a
              name server, and if the query times out, try the next, until
              out of name servers, then repeat trying all the name servers
              until a maximum number of retries are made.)"

unfortunately a name resolution error is not enough to say "try the next one" - it has to actually timeout or fail to connect in order to move to the next.

I don't know how hard or easy this kind of thing is, but perhaps a good long term solution is to just somehow forward NS requests to the native Windows resolver services, which actually do a very good job at this (arguably better than the glibc one).

**Comment by sunilmut** [comment]
@nimbixler - Thanks for the pointer. So, the default DNS resolver on Ubuntu, `network-manager` must have a better logic than `glibc` to do the resolution because obviously it works.

Redirecting the DNS requests to Windows service is not totally out of scope, but the solution from WSL point of view will look a bit different because all the NS resolution in Linux happens in user space, where WSL has no role to play.

**Comment by nimbixler** [comment]
@sunilmut I haven't used desktop Linux in a while, but my guess is that it's providing some sort of caching name service that is being smarter about name resolution on the back end.  It may be worth looking into that approach as well - doing whatever NetworkManager (plus whatever name service) does, perhaps even packaging it up that way.

Just trying to help - you guys did a great thing here, want to make sure it gets used ;-)

Leo Reiter

**Comment by donaldpipowitch** [comment]
I think I made a duplicate of this issue https://github.com/Microsoft/WSL/issues/2884...? Is this the same problem? Are there known workarounds?

**Comment by dannyk81** [comment]
@donaldpipowitch check out this solution https://github.com/Microsoft/WSL/issues/2082#issuecomment-369994557

**Comment by kobenauf** [comment]
It's been almost a year since the last significant comment on this.  Is this still being investigated?  This still seems like a problem.

**Comment by Miles92** [comment]
I also tried to get deadwood running like described here. [issue 2082](https://github.com/Microsoft/WSL/issues/2082#issuecomment-369994557), but it still doesn't work.

Is there any chance of a solution for this?

**Comment by nfekete** [comment]
I'm using this bash script to generate the proper `resolv.conf` based on current windows DNS configuration values, which I run each time the windows DNS settings change (i.e. when connecting to a VPN server) or the `resolv.conf` file gets overwritten by WSL for some reason.
```sh
#!/bin/bash

tmp=`mktemp`
powershell="/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe"
trap ctrlC INT

removeTempFiles() {
    rm -f $tmp
}

ctrlC() {
    echo
    echo "Trapped Ctrl-C, removing temporary files"
    removeTempFiles
    stty sane
}

echo "Current resolv.conf"
echo "-------------------"
cat /etc/resolv.conf

echo
echo "Creating new resolv.conf"
echo "------------------------"

{
    head -1 /etc/resolv.conf | grep '^#.*generated'
    nameservers=$($powershell \
        -Command "Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses" \
        )
    for i in $nameservers; do
        echo nameserver $i
    done
    tail -n+2 /etc/resolv.conf | grep -v '^nameserver'
} | tr -d '\r' | tee $tmp

(set -x; sudo cp -i $tmp /etc/resolv.conf)

removeTempFiles
```

You can mark this script setuid with owner root, so it doesn't ask for password through `sudo` every time.

**Comment by TsukeNoMe** [comment]
I resolved this issue basically by applying the following steps:

1) Locate the VPN adapter over the network settings. To go there:
       a. Go to Settings
       b. Click on Network & Internet
       c. Change adapter options

2) Right click on the VPN virtual network, and click on properties.
3) In my case I selected IPv4, and then click on properties
4) You will see the DNS settings that are set statically

You will need to copy these DNS settings to your /etc/resolv.conf. It is advisable that if the DNS failover is turned on you can add all of the DNS settings from various adapters. A simple script that generates a bash file with the commands for all of the adapters is as the following;


```
function PopulateDNSServers()
{
    $DNSServers = Get-DnsClientServerAddress | Select-Object –ExpandProperty ServerAddresses

    #Create the bash file
    $FileName = 'C:\Users\' + $env:USERNAME + "\WSLDNSFix.bash"
    Write-Host $FileName

    New-Item $FileName -type file

    echo "#!/bin/bash" >> $FileName
    echo "sudo cp /etc/resolv.conf{,.bkp}" >> $FileName
    echo "sudo rm /etc/resolv.conf" >> $FileName
    echo "sudo touch /etc/resolv.conf" >> $FileName

    foreach($serverAddress in $DNSServers)
    {        
        $lineToAdd = "nameserver " + $serverAddress
        $Command = "sudo sh -c ""echo " + $lineToAdd + " >>/etc/resolv.conf"""
        echo $Command >> $FileName
                
    }
}

```
Once these are done, you can restart your WSL, and should be good.

**Comment by deven** [comment]
Here's a shorter bash function to fix the nameserver list with "ed":

`# Fix /etc/resolv.conf for VPN.
vpn() {
   [ -s /run/resolvconf/resolv.conf ] && sudo cp --remove-destination /run/resolvconf/resolv.conf /etc/resolv.conf && sudo ed /etc/resolv.conf >/dev/null 2>&1 <<EOF && diff -c /run/resolvconf/resolv.conf /etc/resolv.conf
g/^nameserver /d
-kx
'xr !/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe -Command "Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses"
'x+1,.s/^/nameserver /
wq
EOF
}`

**Comment by deven** [comment]
Here is a shorter bash function to fix the nameserver list using "ed":
```bash
# Fix /etc/resolv.conf for VPN.
vpn() {
   [ -s /run/resolvconf/resolv.conf ] && sudo cp --remove-destination /run/resolvconf/resolv.conf /etc/resolv.conf && sudo ed /etc/resolv.conf >/dev/null 2>&1 <<EOF && diff -c /run/resolvconf/resolv.conf /etc/resolv.conf
g/^nameserver /d
-kx
'xr !/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe -Command "Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses"
'x+1,.s/^/nameserver /
wq
EOF
}
```
What this does, step-by-step:

1. Verify that "/run/resolvconf/resolv.conf" exists and is not empty.
2. Copy "/run/resolvconf/resolv.conf" to "/etc/resolv.conf", removing the destination file first (in case it's a symbolic link).
3. Run "ed" to update the list of nameservers in "/etc/resolv.conf", discarding the editor output.  The following lines through the EOF marker contain "ed" commands:
  * Delete all lines starting with "nameserver ".
  * Mark the previous line (before the nameserver list) with "x".
  * Insert the output of the specified PowerShell command (fetching the nameservers from Windows) after the marked line.
  * Prefix the new lines with "nameserver ".
  * Save and exit.
4. Display the differences between "/run/resolvconf/resolv.conf" and "/etc/resolv.conf" from the "ed" script.

**Comment by chiamarc** [comment]
Deven,

This works great unless you only have nameserver entries in your resolv.conf.  This was my situation so I had to elide the command to move to the previous line '-' and just make it kx because I was already at the first line.  Is there a more general way to do this?

**Comment by matthiassb** [comment]
@chiamarc I am curious, does the GIST I provided [here](https://github.com/Microsoft/WSL/issues/416#issuecomment-407075002) resolve your DNS issues?

**Comment by deven** [comment]
> Deven,
> 
> This works great unless you only have nameserver entries in your resolv.conf. This was my situation so I had to elide the command to move to the previous line '-' and just make it kx because I was already at the first line. Is there a more general way to do this?

Yes, I came up with a solution and edited my comment above to fix this problem.  The new version should work for your situation.  I updated the explanation as well.

**Comment by chiamarc** [comment]
Thanks @deven !

**Comment by mathrb** [comment]
@deven thanks for the workaround. Any idea with executing the script change the layout of the bash?
![bash_1](https://user-images.githubusercontent.com/4428209/62317455-97476c80-b44e-11e9-8411-d18638c75949.PNG)
![bash_2](https://user-images.githubusercontent.com/4428209/62317461-99113000-b44e-11e9-84b3-819f4ebb6885.PNG)

**Comment by coltenkrauter** [comment]
This simple solution worked for me:
[Fix DNS resolution in WSL2](https://gist.github.com/coltenkrauter/608cfe02319ce60facd76373249b8ca6)

**Comment by sgarcia-dev** [comment]
None of the solutions work for me while running a VPN :( Might it have something to do that running the powershell command, I just get a single "192.168.1.1" address (same as before connecting)? 

Also, when I inspect the connection properties in the command panel -> IPv4 settings -> DNS address, I just get "DNS resolved dynamically", and can't see what the true DNS address is of my currently connected VPN.

Does someone have any ideas?

**Comment by nimbixler** [comment]
I don't know how relevant this is anymore, but this works for me (a bash function that I run from .bashrc):

`dnsfix ()
{
    /mnt/c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2 | grep --color=auto -e '^ [0-9]' | sed 's/^/nameserver/' | sudo tee /etc/resolv.conf > /dev/null
}`

If I ever connect to VPN and have a WSL session open already, and done want to open another one, I just type `dnsfix` in the shell manually.  Otherwise running a new session will do this.  Note that you have to have passwordless sudo setup for your shell user, which I did like this:

`ler@LAPTOP-LEO:~$ cat /etc/sudoers.d/00-ler
ler ALL=(ALL) NOPASSWD: ALL
`

replace ler with whatever your username is for WSL of course.

**Comment by MGlants** [comment]
> I don't know how relevant this is anymore, but this works for me (a bash function that I run from .bashrc):
> 
> `dnsfix () { /mnt/c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2 | grep --color=auto -e '^ [0-9]' | sed 's/^/nameserver/' | sudo tee /etc/resolv.conf > /dev/null }`
> 
> If I ever connect to VPN and have a WSL session open already, and don't want to open another one, I just type `dnsfix` in the shell manually. Otherwise running a new session will do this. Note that you have to have passwordless sudo setup for your shell user, which I did like this:
> 
> `ler@LAPTOP-LEO:~$ cat /etc/sudoers.d/00-ler ler ALL=(ALL) NOPASSWD: ALL `
> 
> replace ler with whatever your username is for WSL of course.
Thank you for your awesome script.
Added fix on DOS CR-LF

`/mnt/c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2 | grep --color=auto -e '^ [0-9]' | sed 's/^/nameserver/' |tr -d '\r' | sudo tee /etc/resolv.conf > /dev/null`

**Comment by gitfool** [comment]
I'm in the same boat as @sgarcia-dev:
> None of the solutions work for me while running a VPN :( Might it have something to do that running the powershell command, I just get a single "192.168.1.1" address (same as before connecting)?

**Comment by MGlants** [comment]
> I'm in the same boat as @sgarcia-dev:
> 
> > None of the solutions work for me while running a VPN :( Might it have something to do that running the powershell command, I just get a single "192.168.1.1" address (same as before connecting)?

Look at mine comment

**Comment by gitfool** [comment]
@MGlants like I said, it doesn't work for me since it only returns `192.168.1.1`, which is already in `/etc/resolv.conf`:
```bash
~$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, remove this line.
nameserver 192.168.1.1
nameserver fec0:0:0:ffff::1
nameserver fec0:0:0:ffff::2

~$ /c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2 | grep --color=auto -e '
^ [0-9]' | sed 's/^/nameserver/' | tr -d '\r'
nameserver 192.168.1.1
```

**Comment by RobberPhex** [comment]
For WSL1 on Win 10 2004, WSL1 works very well with VPN:

Just need to enable `[network]generateResolvConf`.

I.e.
```
robert@OFFICE-LAPTOP:~$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.124.1
nameserver 127.0.0.1
nameserver fec0:0:0:ffff::1
robert@OFFICE-LAPTOP:~$ #disconnect VPN
robert@OFFICE-LAPTOP:~$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.124.1
nameserver fec0:0:0:ffff::1
nameserver fec0:0:0:ffff::2
robert@OFFICE-LAPTOP:~$ #connect VPN
robert@OFFICE-LAPTOP:~$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.124.1
nameserver 127.0.0.1
nameserver fec0:0:0:ffff::1
```

This solution works with MoitonPro and built-in L2TP VPN.

**Comment by zahaim** [comment]
> 
> 
> I don't know how relevant this is anymore, but this works for me (a bash function that I run from .bashrc):
> 
> `dnsfix () { /mnt/c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2 | grep --color=auto -e '^ [0-9]' | sed 's/^/nameserver/' | sudo tee /etc/resolv.conf > /dev/null }`
> 
> If I ever connect to VPN and have a WSL session open already, and don't want to open another one, I just type `dnsfix` in the shell manually. Otherwise running a new session will do this. Note that you have to have passwordless sudo setup for your shell user, which I did like this:
> 
> `ler@LAPTOP-LEO:~$ cat /etc/sudoers.d/00-ler ler ALL=(ALL) NOPASSWD: ALL `
> 
> replace ler with whatever your username is for WSL of course.

I encountered 2 little problems with this script:
- it was unable to work with ipv6 addresses as those use `:` as delimiters
- windows newline characters were taken from the ipconfig.exe command so the /etc/resolv.conf file was no longer understood by linux utilities 
```$ dig
dig: parse of /etc/resolv.conf failed```

I solved those by adding a dash at the end of the `cut` command then I installed `dos2unix` utility to get rid of windows newlines. Working script looks as follows for me now:

```
$ cat dnsfix.sh 
#!/bin/bash
# helper script to update dns setting while on VPN
/mnt/c/Windows/system32/ipconfig.exe /all | grep --color=auto "DNS Servers" | cut -d ":" -f 2- | grep --color=auto -e '^ [0-9]' | sed 's/^/nameserver/' | sudo tee /etc/resolv.conf > /dev/null
sudo dos2unix /run/resolvconf/resolv.conf
```
```

**Comment by matthiassb** [comment]
@zahaim check out my [GIST](https://gist.github.com/matthiassb/9c8162d2564777a70e3ae3cbee7d2e95) It's an Init.d service to keep DNS in sync

To support IPv6 as well as IPv6 remove `-AddressFamily IPv4` from line 22

**Comment by thebiss** [comment]
There's probably a way to do this with 1 line of awk, but here's my fix.  This handles ipv4 AND ipv6 and search parameters.  Don't know if it's i18n aware.

```sh
dnsfix () {     
    local search=$(grep -F "search" /etc/resolv.conf);
    local nameservers=$(/mnt/c/Windows/system32/ipconfig.exe /all | grep -F "DNS Servers" | cut -c 40- | tr -d '\r' | sed 's/^/nameserver /');
    printf '%s on %s\n%s\n%s\n' '# generated by dnsfix' "`date`" "${nameservers}" "${search}" | sudo tee /etc/resolv.conf;
}
```

Which generates the following `/etc/resolv.conf`

```sh
# generated by dnsfix on Thu Apr 16 14:04:20 EDT 2020
nameserver X.X.X.X
nameserver XXXX:6940:6b2e::1
search example.net
```

**Comment by atorr0** [comment]
WSL1: **For some people** may work this simple script, while keeping the file automatically managed by WSL:
```
cat /etc/resolv.conf | sort | sudo tee /etc/resolv.conf
```
Generating the following result:
```
# This file was automatically generated by WSL. To stop automatic generation of this file, remove this line.
nameserver 172.12.34.56
nameserver 8.8.4.4
nameserver 8.8.8.8
search my-home-lan corporate.com
```

**Comment by thebiss** [comment]
Note that sorting may inadvertently place DNS servers added by the VPN lower in the search list, preventing resolution of internal hostnames known only that DNS. 

For example, unlike Google (8.8.8.8), cloudflare (1.1.1.1) would always appear first.

**Comment by davidegiunchi** [comment]
I solved the problem.
In my case the problem was that the OpenVPN configuration that i was using, didn't set any dns resolver, so windows was correctly using google dns, but wsl didn't set any dns.
I solved by modifying the openvpn client configuration by adding this line:
----
dhcp-option DNS 1.1.1.1
----
with this option set, when i connect to openvpn, the dns is getting correctly set even on linux wsl.

**Comment by zahaim** [comment]
> @zahaim check out my [GIST](https://gist.github.com/matthiassb/9c8162d2564777a70e3ae3cbee7d2e95) It's an Init.d service to keep DNS in sync
> 
> To support IPv6 as well as IPv6 remove `-AddressFamily IPv4` from line 22

I was to try it but then I moved to wsl2. Thanks for sharing anyway :)

**Comment by ebc-conscia** [comment]
To neatly and automatically update your resolv.conf file, it is only neccessary to reload Lxss. In the case of split-tunneling, my experience is that nameserver rotation does the trick.
I basically just added this alias to my .bashrc file as a workaround:
`alias reload='powershell.exe -Command "{Restart-Service LxssManager}" && echo "options rotate" | sudo tee -a /run/resolvconf/resolv.conf'`

Of course be sure to check whether the /etc/wsl.conf file is configured correctly:

`[network]` 
`generateResolvConf = true`

**Comment by gusmaogabriels** [comment]
I had the same issue after upgrading from WSL to WSL2. After hours looking up quick fixes online, I accidentally found out that just running WLS2 with Administrator privileges fixed the issue in my case.

**Comment by brelian** [comment]
WSL1 still appears this issue. I solved this problem after adding DNS list from Windows into wsl `/etc/resolve.conf`. Thanks @mo18

**Comment by ebc-conscia** [comment]
Please add the name-servers from the VPN to the top of the resolv.conf file! This would solve the issue for a lot of people. While my file is auto-generated as I (re-)launch WSL, the VPN DNS is put dead last and nothing works.

The blissful ignorance that everyone seems to live in, is that it actually matters to put more than one DNS server in your config. EVERY SINGLE APPLICATION in the world will time out before trying server 2. **The first server in the list matters.** 

Thanks!

**Comment by cifkao** [comment]
> Please add the name-servers from the VPN to the top of the resolv.conf file! This would solve the issue for a lot of people. While my file is auto-generated as I (re-)launch WSL, the VPN DNS is put dead last and nothing works.

Can confirm this. The VPN DNS gets added to resolv.conf, but as the last entry, and it never gets used. It would be great to get this fixed in WSL 1.

**Comment by Christopher-bit** [comment]
> To neatly and automatically update your resolv.conf file, it is only neccessary to reload Lxss. In the case of split-tunneling, my experience is that nameserver rotation does the trick.
> I basically just added this alias to my .bashrc file as a workaround:
> `alias reload='powershell.exe -Command "{Restart-Service LxssManager}" && echo "options rotate" | sudo tee -a /run/resolvconf/resolv.conf'`
> 
> Of course be sure to check whether the /etc/wsl.conf file is configured correctly:
> 
> `[network]`
> `generateResolvConf = true`

thanks a lot! I came up with this problem too and I'm counting it on my wrongly editing of the wsl.conf.   Anyway I fixed it by using your experience. Thx a lot!

**Comment by Spongman** [comment]
this is still not fixed in 1904

**Comment by funmonty** [comment]
I had the same issue using WSL2. Tried a lot of things, everytime the /etc/resolv.conf was getting updated automatically. Finally tried switching back to WSL1, the problem was solved.

**Comment by ericis** [comment]
Thanks @Christopher-bit . That's a partial fix for me. I cannot run `sudo apt-get update` when using NordVPN. However, I couldn't  even `nslookup google.com` until I tried your workaround. This problem is related to #4648 (already mentioned above).

**Comment by newcarrotgames** [comment]
soo.. for me this was caused by me starting my VPN before starting WSL. I fixed it by disconnecting from the VPN, restarting WSL (wsl -t <distro>), and then reconnecting to the VPN.

**Comment by nrcmkoh** [comment]
Not sure if this will help anyone, but throwing in my 2 cents.

Sorting the private nameserver to the top of the list allowed me to reach hosts in my private network, but then I could no longer reach public ones. I wanted to access both without have to connect/disconnect from VPN, and adding `options rotate` to the top of the `/etc/resolve.conf` allowed me to do that.

**Comment by kevinlrak** [comment]
Update to the `dnsfix` to use PowerShell:
```bash
dnsfix () {
  pwsh.exe -Command "Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses" | sed 's/^/nameserver /' | sudo tee /etc/resolv.conf > /dev/null;
}
```

**Comment by dmslowmo** [comment]
I am using WSL2 and Cisco AnyConnect VPN. What solved this for me is:
1. Open AnyConnect "Advanced Window". For me, it is clicking the gear icon.
2. Go to VPN -> Preferences
3. Tick the checkbox "Allow local (LAN) access when using VPN (if configured)

The default content of /etc/resolv.conf for me is "nameserver 192.168.*.*"

Hope this helps someone.

**Comment by piymis** [comment]
> I am using WSL2 and Cisco AnyConnect VPN. What solved this for me is:
> 
> 1. Open AnyConnect "Advanced Window". For me, it is clicking the gear icon.
> 2. Go to VPN -> Preferences
> 3. Tick the checkbox "Allow local (LAN) access when using VPN (if configured)
> 
> The default content of /etc/resolv.conf for me is "nameserver 192.168._._"
> 
> Hope this helps someone.

Thank you so much. This worked for me on WSL2 + Cisco Anyconnect VPN.

**Comment by khalid-kherrazi** [comment]
> > I am using WSL2 and Cisco AnyConnect VPN. What solved this for me is:
> > 
> > 1. Open AnyConnect "Advanced Window". For me, it is clicking the gear icon.
> > 2. Go to VPN -> Preferences
> > 3. Tick the checkbox "Allow local (LAN) access when using VPN (if configured)
> > 
> > The default content of /etc/resolv.conf for me is "nameserver 192.168._._"
> > Hope this helps someone.
> 
> Thank you so much. This worked for me on WSL2 + Cisco Anyconnect VPN.

Thanks!! it work for me : WSL1 + Cisco Anyconnect

**Comment by ileathan** [comment]
Only way I got this too work is to add 

`generateResolvConf = false` to `wsl.conf` but im pretty sure something will break down the line from this so if anyone knows let me know because so far its all working great. Altho  like above I had to whitelist my network `nordvpn whitelist add subnet 192.168.2.0/24`

No matter what I was doing before unless i used scripts to edit and re-edit /etc/resolv.conf the internet on vpn always ended up breaking. But if my fix breaks something else down the line ill take more time to read about this issue, right now seems like theres a tun of fixes and most dont really work for me.

**Comment by pedroaugustosmribeiro** [comment]
Add and hybrid solution on wsl.conf for the automatic generation of /etc/resolv.conf with a list of parameters (nameservers, search domains, options like rotate) to be appended to the end of the autogenerated section.

**Comment by donatelloOo** [comment]
**Please note an important fact:**

In my case the DNS servers from VPN connections were not added to /etc/resolv.conf when the other network was using IPv6.
Simply **disabling IPv6 on the ethernet/wifi adapter** allowed to make the VPN connection to work properly.

![image](https://user-images.githubusercontent.com/29257279/101763362-facfb880-3ade-11eb-9eea-4b08633a18bc.png)

Now the nameservers are correctly added when VPN connection is enabled, and removed when VPN is disabled.

With VPN connected:
```
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.1.1
nameserver xx.xx.xx.x1 # obfuscated company dns
nameserver xx.xx.xx.x1 # obfuscated company dns
search home
```

With VPN disconnected:
```
$ cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 192.168.1.1
nameserver fec0:0:0:ffff::1
nameserver fec0:0:0:ffff::2
search home
```

Note that I don't know where does the `fec0:0:0:ffff::1` and `fec0:0:0:ffff::2` IPv6 addresses come from...

**Comment by ebc-conscia** [comment]
I have added the following to my `~/.bashrc` file. This is a consistent fix that takes care of the interface priority problem, via an auto-elevating PowerShell script and solves the DNS issue with VPN. Note that you have to create the script as described in the comments inside here and change interface name according to your provider (in my case AnyConnect).

When you connect to VPN via AnyConnect or another provider simply write `vpn ` in WSL/WSL2 prompt and you'll be asked to type the sudo password and approve the PS elevation.

Hope this helps you guys, but my gosh how hacky is this? 

I really hope Microsoft can deliver a more robust fix via Windows Update, since now everyone is WFH, that also means being connected to VPN most of the time.

```bash
#run to fix connection / nameserver issues when connected to VPN
vpn () {
#Write dns servers from VPN interface to /etc/resolv.conf
    powershell.exe -Command 'Get-NetAdapter -InterfaceDescription "Cisco AnyConnect*" | Get-DnsClientServerAddress | Select -ExpandProperty ServerAddresses' | awk 'BEGIN { print "# Generated by vpn fix func on", strftime("%c"); print } { print "nameserver", $1 }' | tr -d '\r' | sudo tee /etc/resolv.conf
#                                                         ## ^VPN^PROVIDER^INTERFACE^NAME^ ##
#Script to set interface metric correctly, since admin is required, some extra code to elevate rights.
    powershell.exe -File 'C:\\Users\\USERNAME\\dev\\Scripts\\vpn.ps1'
#                        ^YOUR SCRIPT LOCATION IN WINDOWS FORMAT^
#
#THE SCRIPT SHOULD CONTAIN:
#
#if (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
#    Start-Process PowerShell -Verb RunAs "-NoProfile -ExecutionPolicy Bypass -Command `"cd '$pwd'; & '$PSCommandPath';`"";
#    exit;
#}
#Get-NetAdapter | Where-Object {$_.InterfaceDescription -Match "Cisco AnyConnect"} | Set-NetIPInterface -InterfaceMetric 6000
#                                                    ## ^VPN^PROVIDER^INTERFACE^NAME^ ##
#
}
```

Cheers

**Comment by nivekastoreth** [comment]
thank you so much @donatelloOo, [your post](https://github.com/microsoft/WSL/issues/1350#issuecomment-742454940) is the only thing that has gotten my AnyConnect working with WSL.

One note however, I've noticed I have to disable IPv6 on the WSL adapter (and possibly the Hyper-V bridge as well) each time I restart and before I open WSL for the first time.

**Comment by rg1235** [comment]
Still facing issues with DNS resolution in windows 20H2 build with Ubuntu 20.04 on WSL2.
Hostname resolution works sometimes and sometimes it doesn't , not at all reliable . I am also using Vcxrvs X server to run linux GUI apps, if I set `generateResolvConf = false` and add companies DNS at the starting of /etc/resolv.conf then on starting apps for GUI it complains as /etc/resolv.conf only takes into consideration the first nameserver available.
Please need a fix

**Comment by trallnag** [comment]
@rg1235, you can use DNSmasq. Then you point to it in your resolv.conf and the service then takes over sending resolv requests to different DNS. You will have to set up a separate config for DNSmasq as well containing the different DNS server you want to use.

I am using it to deal with the split DNS situation in my company and it works very well. Just remember to start it up every time when the WSL2 box starts (add start code to your profile for example).

**Comment by rg1235** [comment]
> @rg1235, you can use DNSmasq. Then you point to it in your resolv.conf and the service then takes over sending resolv requests to different DNS. You will have to set up a separate config for DNSmasq as well containing the different DNS server you want to use.
> 
> I am using it to deal with the split DNS situation in my company and it works very well. Just remember to start it up every time when the WSL2 box starts (add start code to your profile for example).

@trallnag thanks for the reply just want to confirm a few things 
 Do you have X server configured? , I currently use gui for linux apps and WSL2 automatically resolves the address of the display based upon entries in resolv.conf ,so correct me if I am wrong if I add DNSmasq to resolv.conf (as the first entry) will gui apps find the running X server itself?
And it will be helpful if you can share your dnsmasq configuration file sample , Thanks again

**Comment by trallnag** [comment]
@rg1235

```
port=53
no-resolv
server=/<intranet-top-level-domain>/<internal-dns>
server=/<intranet-top-level-domain>/<internal-dns>
server=8.8.8.8
```

Regarding X server: No, I don't use it atm

**Comment by jonigl** [comment]
> **Please note an important fact, and workaround:**
> 
> The DNS servers from VPN connections are not added to `/etc/resolv.conf` when another network adapter is using **IPv6 DNS servers**, which seems to cause kind of conflict (additional IPv4 DNS servers are discarded).
> 
> Check the adapters that use IPv6 DNS servers :
> 
> ```powershell
> Get-DnsClientServerAddress -AddressFamily IPv6 | Where-Object ServerAddresses -NE "{}" | Select-Object -ExpandProperty InterfaceAlias
> ```
> 
> Get their related adapter binding
> 
> ```powershell
> Get-NetAdapterBinding -ComponentID ms_tcpip6 | Where-Object Name -In (Get-DnsClientServerAddress -AddressFamily IPv6 | Where-Object ServerAddresses -NE "{}" | Select-Object -ExpandProperty InterfaceAlias)
> ```
> 
> Disable IPv6 for each adapter binding (or directly for all adapters) using Powershell **with administrator privileges**:
> 
> ```powershell
> Disable-NetAdapterBinding -Name "Wi-Fi" -ComponentID ms_tcpip6 -PassThru
> Disable-NetAdapterBinding -Name "Network Bridge" -ComponentID ms_tcpip6 -PassThru
> ...
> Disable-NetAdapterBinding -Name "*" -ComponentID ms_tcpip6 -PassThru
> ```
> 
> Alternatively, simply **disable IPv6 on the ethernet/wifi adapter** using Windows UI:
> 
> ![image](https://user-images.githubusercontent.com/29257279/101763362-facfb880-3ade-11eb-9eea-4b08633a18bc.png)
> 
> Now the nameservers are correctly added when VPN connection is enabled, and removed when VPN is disabled.
> 
> With VPN connected:
> 
> ```
> $ cat /etc/resolv.conf
> # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
> # [network]
> # generateResolvConf = false
> nameserver 192.168.1.1
> nameserver xx.xx.xx.x1 # obfuscated company dns
> nameserver xx.xx.xx.x1 # obfuscated company dns
> search home
> ```
> 
> With VPN disconnected:
> 
> ```
> $ cat /etc/resolv.conf
> # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
> # [network]
> # generateResolvConf = false
> nameserver 192.168.1.1
> search home
> ```

Thank you @donatelloOo I was experiencing network problems when using Hamachi VPN. 
The Hamachi adapter hasd IPv6 enabled and by disabling it I was able to resolve the issue!!! 😄  🎉

**Comment by duyn9uyen** [comment]
@funmonty, I also had to switched back to WSL1. WSL2's resolv.conf kept getting updated automatically and stopped my connectivity.

**Comment by cod3monk3y** [comment]
> 
> 
> I have added the following to my `~/.bashrc` file. This is a consistent fix that takes care of the interface priority problem, via an auto-elevating PowerShell script and solves the DNS issue with VPN. Note that you have to create the script as described in the comments inside here and change interface name according to your provider (in my case AnyConnect).
> 
> When you connect to VPN via AnyConnect or another provider simply write `vpn ` in WSL/WSL2 prompt and you'll be asked to type the sudo password and approve the PS elevation.
> 
> Hope this helps you guys, but my gosh how hacky is this?
> 
> I really hope Microsoft can deliver a more robust fix via Windows Update, since now everyone is WFH, that also means being connected to VPN most of the time.
> 
> ```shell
> #run to fix connection / nameserver issues when connected to VPN
> vpn () {
> #Write dns servers from VPN interface to /etc/resolv.conf
>     powershell.exe -Command 'Get-NetAdapter -InterfaceDescription "Cisco AnyConnect*" | Get-DnsClientServerAddress | Select -ExpandProperty ServerAddresses' | awk 'BEGIN { print "# Generated by vpn fix func on", strftime("%c"); print } { print "nameserver", $1 }' | tr -d '\r' | sudo tee /etc/resolv.conf
> #                                                         ## ^VPN^PROVIDER^INTERFACE^NAME^ ##
> #Script to set interface metric correctly, since admin is required, some extra code to elevate rights.
>     powershell.exe -File 'C:\\Users\\USERNAME\\dev\\Scripts\\vpn.ps1'
> #                        ^YOUR SCRIPT LOCATION IN WINDOWS FORMAT^
> #
> #THE SCRIPT SHOULD CONTAIN:
> #
> #if (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
> #    Start-Process PowerShell -Verb RunAs "-NoProfile -ExecutionPolicy Bypass -Command `"cd '$pwd'; & '$PSCommandPath';`"";
> #    exit;
> #}
> #Get-NetAdapter | Where-Object {$_.InterfaceDescription -Match "Cisco AnyConnect"} | Set-NetIPInterface -InterfaceMetric 6000
> #                                                    ## ^VPN^PROVIDER^INTERFACE^NAME^ ##
> #
> }
> ```
> 
> I got some inspiration from different posts on this site, sorry for not crediting those, I don't actually remember the who and where.
> 
> Cheers

After trying *all* the other solutions on the *entire* internet, this was the only option that worked for me. I augmented this with a `notvpn` function as well to restore the nameservers to Google's `8.8.8.8` and `8.8.4.4` to have the ability to switch DNS manually when I connect/disconnect VPN. Now when I connect VPN i run `vpn` and when I disconnect VPN I run `notvpn`. 

Thanks for this. The issue has been plaguing me for weeks.

Here's my hacky `notvpn` function

```
notvpn () {
    echo "# generated by ~/.bashrc notvpn () "
    echo "nameserver 8.8.8.8" | tr -d '\r' | sudo tee /etc/resolv.conf
    echo "nameserver 8.8.4.4" | tr -d '\r' | sudo tee /etc/resolv.conf
}
```

**Comment by davidcantrell** [comment]
The solution from @cod3monk3y worked when added to .bash_profile. Thank you for this workaround.

Microsoft: Totally unacceptable to have these issues. After spending an afternoon moving to WSL2 with Docker for Windows, I am not amused after finding DNS resolutions didn't work the next day.

echo "# generated by ~/.bashrc notvpn () "
echo "nameserver 8.8.8.8" | tr -d '\r' | sudo tee /etc/resolv.conf
echo "nameserver 8.8.4.4" | tr -d '\r' | sudo tee /etc/resolv.conf

**Comment by cod3monk3y** [comment]
> The solution from @cod3monk3y worked when added to .bash_profile. Thank you for this workaround.

Credit for this should really go to https://github.com/ebc-conscia . I really just added the opposite method to their `vpn()` solution/workaround. 

I'll note that sometimes this still doesn't work on the first attempt for me. I still sometimes have to shut down WSL, Cisco AnyConnect, then reconnect VPN (Cisco), then reopen WSL, then type `vpn`. When I get ony one DNS entry (primary), the `vpn` settings don't work. But if I get both primary and secondary DNS entries, the networking will work correctly. 

Not sure if this problem is originating from Cisco AnyConnect or not. I feel like it could be Cisco not populating the secondary DNS correctly.

And, I'm in the habit of *always* trying `host google.com` when I switch between `vpn` and `notvpn` modes.

But I'm usually able to get into a state where networking is working. It's just a matter of "try again", "how about now?" iterating.

**Comment by cazzoo** [comment]
For those still looking for a solution to get it working, I've been applying https://github.com/microsoft/WSL/issues/5068#issuecomment-839756030 and it's working fine for me.

**Comment by spencerharmon** [comment]
Could I suggest that microsoft gives the ability to configure dns forwarding on the resolver that gets configured when resolv.conf is generated? I think everyone is expecting that WSL2 will automatically use the DNS configuration of its host without having to manually switch static DNS servers. This could be easily accomplished with forwarding.

**Comment by donatelloOo** [comment]
> 
> 
> Could I suggest that microsoft gives the ability to configure dns forwarding.

Actually, this is the default case. and the goal of `generateResolveConfg`
https://docs.microsoft.com/en-us/windows/wsl/wsl-config#network

but it may fail due with VPN or ipv6 servers... since years.

**Comment by spencerharmon** [comment]
I'm saying that microsoft should provide a proper DNS forwarder as it's defined in RFC2308, like the feature in bind, unbound, dnsmasq, et c., at the IP address that's set as the gateway and as the nameserver when starting a WSL2 instance. Changing resolv.conf more than one time is specifically what I'm saying should not be done. It should be set once to this hypothetical forwarder, which should have the capability to dynamically forward requests (or not to do so, assuming the forwarder itself is fully and independently configurable) according to the DNS configuration of the host, which could be changed at any time due to moving between wireless networks or connecting to a VPN. I don't think I'm mistaken when I say that microsoft has never provided a DNS forwarder with WSL2 on win10.

**Comment by davidcantrell** [comment]
Having a new IP address assigned to WSL2 every time it is launched creates additional work when DNS and virtual environments are used to access web applications for testing and development. 

As a result of this issue, I have abandoned WSL2 and moved back to Ubuntu on VirtualBox. This has eliminated many headaches. I don't understand why this issue wasn't fixed before but can't allow any more impacts to my projects.

**Comment by Kraego** [comment]
Run into the same issues.

Setup:
* Windows 20H2(Build 19042.1110)
* WSL2 Debian
* VPN: Cisco Any Connect

I've created a workaround [see here](https://stackoverflow.com/a/68573952/11473934) which is basically a excerpt of the comments of this issue.

As @spencerharmon stated: **Microsoft realy should implement a proper DNS forwarder** that resloves the issues when working with VPN, it's a standard scenario to have a VPN especially in the current situation.

**Comment by qci-amos** [comment]
I've spent a lot of time puzzling over this on my brand new laptop / installation only to discover that WSL1 was installed by default (❗❗), not WSL2. I can confirm that a fresh WSL1 doesn't work (I wrote a custom python script to workaround it), but WSL2 does (both "Ubuntu", using built-in vpn).

**Comment by teesamuel** [comment]
> I'm using this bash script to generate the proper `resolv.conf` based on current windows DNS configuration values, which I run each time the windows DNS settings change (i.e. when connecting to a VPN server) or the `resolv.conf` file gets overwritten by WSL for some reason.
> 
> ```shell
> #!/bin/bash
> 
> tmp=`mktemp`
> powershell="/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe"
> trap ctrlC INT
> 
> removeTempFiles() {
>     rm -f $tmp
> }
> 
> ctrlC() {
>     echo
>     echo "Trapped Ctrl-C, removing temporary files"
>     removeTempFiles
>     stty sane
> }
> 
> echo "Current resolv.conf"
> echo "-------------------"
> cat /etc/resolv.conf
> 
> echo
> echo "Creating new resolv.conf"
> echo "------------------------"
> 
> {
>     head -1 /etc/resolv.conf | grep '^#.*generated'
>     nameservers=$($powershell \
>         -Command "Get-DnsClientServerAddress -AddressFamily ipv4 | Select-Object -ExpandProperty ServerAddresses" \
>         )
>     for i in $nameservers; do
>         echo nameserver $i
>     done
>     tail -n+2 /etc/resolv.conf | grep -v '^nameserver'
> } | tr -d '\r' | tee $tmp
> 
> (set -x; sudo cp -i $tmp /etc/resolv.conf)
> 
> removeTempFiles
> ```
> 
> You can mark this script setuid with owner root, so it doesn't ask for password through `sudo` every time.

After running this bash script, you can make the file resolv.conf immutable by running `sudo chattr +i /etc/resolv.conf` so that  wsl will not overwrite your settings.

to undo the command, run
`sudo chattr -i /etc/resolv.conf`

**Comment by rubin55** [comment]
Just a note to tell readers here I wrote a nice fix for this that is quite a bit simpler + less error-prone than some of the solutions I've found floating around, using Windows Task Scheduler. In short, it's a powershell script that can write a correct resolv.conf based on InterfaceMetric order (so just make sure your VPN adapter has a lower (which means higher priority) InterfaceMetric than your WiFi or Ethernet card). Have a look here:

https://github.com/Microsoft/WSL/issues/2884#issuecomment-928299305

**Comment by OmriSama** [comment]
Has anyone been able to get this to work with PaloAlto's GlobalProtect? We keep having issues and it's so frustrating.

**Comment by boan-jfm** [comment]
Anyone knows about the `../run/resolvconf/resolv.conf`?

https://docs.microsoft.com/da-dk/windows/wsl/troubleshooting#bash-loses-network-connectivity-once-connected-to-a-vpn

**Comment by zenithstorm** [comment]
I've been trying all sorts of suggestions with the /etc/resolv.conf file, but I still can't seem to connect to my company's VPN. I can see the connections when I run `ipconfig` on Windows, but not when I run `ifconfig` in WSL.

**Comment by trallnag** [comment]
@zenithstorm, have you tried dnsmasq approach?

**Comment by zenithstorm** [comment]
@trallnag Thank you for the suggestion! I've been trying it for the past half hour or so, but no luck so far. I will keep fiddling with it.

**Comment by sanarena** [comment]
In my case, i set network interface metric to 6000 and both vpn and internet within wsl is now working:
`Get-NetAdapter | Where-Object {$_.InterfaceDescription -Match "Cisco AnyConnect"} | Set-NetIPInterface -InterfaceMetric **6000**`

**Comment by spookyrecharge** [comment]
Don't forget to make **/etc/resolv.conf** file immutable after editing to avoid Windows overwrite it

`chattr +i /etc/resolv.conf`

I also was struggling with using wireguard tunnel
It was important to me since I'm using DoH on MikroTik wireguard server

**Comment by chrisdlangton** [comment]
This comment has the solution that worked best for me on WSL2
https://github.com/microsoft/WSL/issues/2884#issuecomment-932483876
Hope [wsl-vpnkit](https://github.com/sakai135/wsl-vpnkit) helps you too

**Comment by marwin1991** [comment]
1. Find out nameserver with windows powershell (during VPN Session and without) using `nslookup`
2. USe `sudo touch /etc/wsl.conf` and `sudo vim /etc/wsl.conf` to add:
```
[network]                                                                        
generateResolvConf = false
```
3. Restart wsl (Windows powershell) using `wsl --shutdown`
4. Open WSL and remove using `rm -f /etc/resolv.conf`
5. Add new file `sudo touch /etc/resolv.conf` and `sudo vim /etc/resolv.conf` with:
```
nameserver X.X.X.X

nameserver Y.Y.Y.Y
```
6. Restart wsl (Windows powershell) using `wsl --shutdown`
7. Open WSL and remove using `wget google.com` and test some you corporate domain.

**Comment by 2-X** [comment]
don't forget to make sure your VPN's `nameserver` comes **first** in your `/etc/resolv.conf` file
i was losing my sanity to this issue lol
luckily this seems to have done the trick for me

**Comment by EdwardCooke** [comment]
Here's my solution, it takes what @cod3monk3y did, and makes it automatic it so you don't have to run anything manually anymore. It's geared towards the Cisco AnyConnect client, so if you you're using something different, you would just need to figure out the events in the event log to trigger off of.

https://www.frakkingsweet.com/automatic-dns-configuration-with-wsl-and-anyconnect-client/

**Comment by Kraego** [comment]
> Here's my solution, it takes what @cod3monk3y did, and makes it automatic it so you don't have to run anything manually anymore. It's geared towards the Cisco AnyConnect client, so if you you're using something different, you would just need to figure out the events in the event log to trigger off of.
> 
> https://www.frakkingsweet.com/automatic-dns-configuration-with-wsl-and-anyconnect-client/

Works like a charm, thanks.

**Comment by githubuser6000** [comment]
Still facing this issue in WSL2. When using ProtonVPN. 

Why is this issue not resolved since 2017? This makes WSL un-useable in corporate envs and home envs, without all these hacky workarounds. Is MS assuming that the workaround is good enough?

**Comment by sanarena** [comment]
isn’t [wsl-vpnkit](https://github.com/sakai135/wsl-vpnkit) working for you?

**Comment by githubuser6000** [comment]
A reboot fixed it 🤷‍♂️

**Comment by ajgrier** [comment]
All the presented "solutions" seem to involve modifying resolv.conf. The default(?) WSL setup has resolv.conf pointing to an internal IP address, and the same address is also used as default route. The failure seems to be that WSL (which handles this internal routing and DNS server / proxy) is not tracking resolver changes on the windows side. All the modifications to resolv.conf are just workarounds for WSL's failure, not fixes to WSL.

Why are windows applications able to track resolver changes, but WSL isn't?

**Comment by Jan-Pleva** [comment]
Hi, is the VPN in development plan? We are waiting for this.

**Comment by exeral** [comment]
unlike https://github.com/microsoft/WSL/issues/2884#issuecomment-932483876 which push DNS from Windows to WSL
I have made a script for the other way: pull DNS from WSL.
https://gist.github.com/exeral/87c792d20262026318661fdc03ea8807

hope it helps ;)

**Comment by Jan-Pleva** [comment]
Thanks. Still I would need something more instant for enterprise environment.

**Comment by saraiva82** [comment]
all I have to say is that after many hours wasted while working off the VPN I decided to just go back to my VM 
afterall 
![image](https://github.com/microsoft/WSL/assets/24394475/1164db52-01ea-4de7-8063-739d065f3e8e)
They know!
i just dont think they care lol

**Comment by saraiva82** [comment]
I just went back to my vm after many hours wasted troubleshooting this.
Afterall.... 
![image](https://github.com/microsoft/WSL/assets/24394475/ba6541a4-9fa4-4c72-bd92-a2752586e9c6)

They know and dont care lol

**Comment by ccbond** [comment]
> 在浪费了很多时间进行故障排除后，我刚刚回到我的虚拟机。 毕竟.... ![图像](https://private-user-images.githubusercontent.com/24394475/292315652-ba6541a4-9fa4-4c72-bd92-a2752586e9c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MjI3MTksIm5iZiI6MTcwNDgyMjQxOSwicGF0aCI6Ii8yNDM5NDQ3NS8yOTIzMTU2NTItYmE2NTQxYTQtOWZhNC00YzcyLWJkOTItYTI3NTI1ODZlOWM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA5VDE3NDY1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiODNmYTNhYjBiZWQxYjY4YTlhMmNlNzUzOTc5NWUyZjVjOGNiNTc1MGQwMTQ3MDE3ZGRjMjkyNGNlODNhYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._weV16sMLpEys6X1VgYT2kwmevKhJ9hLBc80rYy36yA)
> 
> 他们知道但不在乎哈哈

fu*k

**Comment by danielwagn3r** [comment]
For me the new WSL2dns tunneling feature solved the problem, see https://learn.microsoft.com/en-us/windows/wsl/wsl-config#main-wsl-settings

Add the following to `%UserProfile%\.wslconfig` to enable the feature:

```ini
# Settings apply across all Linux distros running on WSL 2
[wsl2]
# Changes how DNS requests are proxied from WSL to Windows
dnsTunneling=true
```

**Comment by selmison** [comment]
This works in my case: https://github.com/microsoft/WSL/issues/4246#issuecomment-1726524286

**Comment by jaredbrogan** [comment]
> For me the new WSL2dns tunneling feature solved the problem, see https://learn.microsoft.com/en-us/windows/wsl/wsl-config#main-wsl-settings
> 
> Add the following to `%UserProfile%\.wslconfig` to enable the feature:
> 
> ```ini
> # Settings apply across all Linux distros running on WSL 2
> [wsl2]
> # Changes how DNS requests are proxied from WSL to Windows
> dnsTunneling=true
> ```

Thanks for pointing that out. Unfortunately, it's only available in Windows 11. 😢
![image](https://github.com/microsoft/WSL/assets/42228088/95b7d588-6312-4377-98f9-670a8aaa6fe1)

**Comment by dendr203** [comment]
> In my case, i set VPN network interface metric to 6000 and both vpn and internet within wsl is now working: `Get-NetAdapter | Where-Object {$_.InterfaceDescription -Match "Cisco AnyConnect"} | Set-NetIPInterface -InterfaceMetric 6000` Cisco AnyConnect mentioned in command above is my VPN. yours could be different. along with the metric number.
> 
> and then used @cod3monk3y solution above for fixing DNS.

This was the only think that have worked for me and resolved the issue.

**Comment by TuanChauKMS** [comment]
Guys, are there any plans to completely fix this issue in the future?
Thanks all :)

**Comment by fedorzh** [comment]
> For me the new WSL2dns tunneling feature solved the problem, see https://learn.microsoft.com/en-us/windows/wsl/wsl-config#main-wsl-settings
> 
> Add the following to `%UserProfile%\.wslconfig` to enable the feature:
> 
> ```ini
> # Settings apply across all Linux distros running on WSL 2
> [wsl2]
> # Changes how DNS requests are proxied from WSL to Windows
> dnsTunneling=true
> ```

I am actually on Windows 11 23H2 and this still doesn't work to access corporate resources. I have internet access even on VPN but the internal corp things can't be accessed.

**Comment by OneBlue** [comment]
Hi ! Please try the [latest networking features](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling) that we've added in WSL. Those should greatly improve compatibility with VPN's.

If the issue still remains, please reopen this issue.

**Comment by rgmz** [comment]
> Hi ! Please try the [latest networking features](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling) that we've added in WSL. Those should greatly improve compatibility with VPN's.

That link doesn't state it, but the linked page ["Advanced settings configuration in WSL"](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#main-wsl-settings) makes it clear that these settings require Windows 11.

@OneBlue do you know if there are plans to support `dnsTunneling` and `autoProxy` in Windows 10?

**Comment by froh** [comment]
unfortunately I'm stuck with corporate windows 10 for another year or two

sigh...

**Comment by terlar** [comment]
Don't be too sad about it, it is not sure that it will fix the issues anyways. I am currently using Citrix Secure Access and I still have to use wsl-vpnkit to get connectivity when on the VPN.

```
[wsl2]
networkingMode=mirrored
firewall=true
dnsTunneling=true
autoProxy=true
```

**Comment by titanofold** [comment]
`dnsTunneling=true` does help resolve the name to an IP address, but I still can't connect to internal resources. I've tried with and without `networkingMode=mirrored`.

**Comment by zeel01** [comment]
> Hi ! Please try the [latest networking features](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling) that we've added in WSL. Those should greatly improve compatibility with VPN's.
> 
> If the issue still remains, please reopen this issue.

What about WSL 1? Due to the filesystem performance issue in WSL 2, it remains unusable for my needs. If I'm understanding this new feature correctly, it's only available to WSL 2.

**Comment by jxu** [comment]
Networking mode as mirrored and DNS Tunneling enabled works for me.
This is in WSL Settings in the Start Menu, under Networking.
